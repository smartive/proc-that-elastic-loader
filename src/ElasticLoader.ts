import { Client, ClientOptions } from '@elastic/elasticsearch';
import { Loader } from 'proc-that';
import { EMPTY, from, Observable, throwError } from 'rxjs';
import { Buffer } from './helpers/Buffer';

class NoIdProvidedError extends Error {
  constructor(public object: any) {
    super('No id provided by object');
  }
}

export class ElasticLoader implements Loader {
  private esClient: Client;
  private buffer: Buffer<any>;

  constructor(
    config: ClientOptions,
    private index: string,
    private predicate: (obj: any) => boolean = () => true,
    private idSelector: (obj: any) => any = (o) => o.id,
    bufferSize = 10
  ) {
    const esConfig = JSON.parse(JSON.stringify(config)) as ClientOptions;
    if (!esConfig.requestTimeout) {
      // set requestTimeout to 5min.
      // reason: when you shoot many index requests to the esClient, elasticsearch buffers your requests.
      // after the default timeout of 30s is exceeded, you receive a TimeoutError and the whole index process fails.
      esConfig.requestTimeout = 1000 * 60 * 5;
    }

    this.buffer = new Buffer(bufferSize);
    this.esClient = new Client(esConfig);
  }

  public write(object: any): Observable<any> {
    if (!this.predicate(object)) {
      return EMPTY;
    }

    const id = this.idSelector(object);

    if (id === null || id === undefined) {
      return throwError(() => new NoIdProvidedError(object));
    }

    const promise = this.buffer.read().then(() =>
      this.esClient.index({
        id,
        index: this.index,
        body: object,
      })
    );

    this.buffer.write(object);

    return from(promise);
  }
}
