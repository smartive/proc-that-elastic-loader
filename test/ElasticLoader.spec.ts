import { of } from 'rxjs';

import { ElasticLoader } from '../src/ElasticLoader';

describe('ElasticLoader', () => {
  let loader: ElasticLoader;
  let client: any;

  beforeEach(() => {
    client = {
      index: (o) => of(o),
    };

    client.index = jest.fn(client.index);
  });

  it('should resolve on correct usage', (done) => {
    loader = new ElasticLoader({ nodes: ['http://localhost:9200'] }, 'testIndex');
    (loader as any).esClient = client;

    loader.write({ id: 1, text: 'test' }).subscribe({ error: done, complete: done });
  });

  it('should use correct index', (done) => {
    loader = new ElasticLoader({ nodes: ['http://localhost:9200'] }, 'testIndex');
    (loader as any).esClient = client;

    loader.write({ id: 1, text: 'test' }).subscribe({
      error: done,
      complete: () => {
        try {
          expect(client.index.mock.calls.length).toBe(1);
          expect(client.index.mock.calls[0][0]).toMatchSnapshot();
          done();
        } catch (e) {
          done(e);
        }
      },
    });
  });

  it('should use correct type', (done) => {
    loader = new ElasticLoader({ nodes: ['http://localhost:9200'] }, 'testIndex');
    (loader as any).esClient = client;

    loader.write({ id: 1, text: 'test' }).subscribe({
      error: done,
      complete: () => {
        try {
          expect(client.index.mock.calls.length).toBe(1);
          expect(client.index.mock.calls[0][0]).toMatchSnapshot();
          done();
        } catch (e) {
          done(e);
        }
      },
    });
  });

  it('should use correct id', (done) => {
    loader = new ElasticLoader({ nodes: ['http://localhost:9200'] }, 'testIndex');
    (loader as any).esClient = client;

    loader.write({ id: 1, text: 'test' }).subscribe({
      error: done,
      complete: () => {
        try {
          expect(client.index.mock.calls.length).toBe(1);
          expect(client.index.mock.calls[0][0]).toMatchSnapshot();
          done();
        } catch (e) {
          done(e);
        }
      },
    });
  });

  it('should use correct body', (done) => {
    loader = new ElasticLoader({ nodes: ['http://localhost:9200'] }, 'testIndex');
    (loader as any).esClient = client;

    loader.write({ id: 1, text: 'test' }).subscribe({
      error: done,
      complete: () => {
        try {
          expect(client.index.mock.calls.length).toBe(1);
          expect(client.index.mock.calls[0][0]).toMatchSnapshot();
          done();
        } catch (e) {
          done(e);
        }
      },
    });
  });

  it('should be called on truthly predicate', (done) => {
    loader = new ElasticLoader({ nodes: ['http://localhost:9200'] }, 'testIndex', (o) => o.text === 'test');
    (loader as any).esClient = client;

    loader.write({ id: 1, text: 'test' }).subscribe({
      error: done,
      complete: () => {
        try {
          expect(client.index.mock.calls.length).toBe(1);
          done();
        } catch (e) {
          done(e);
        }
      },
    });
  });

  it('should not called on falsly predicate', (done) => {
    loader = new ElasticLoader({ nodes: ['http://localhost:9200'] }, 'testIndex', (o) => o.text !== 'test');
    (loader as any).esClient = client;

    loader.write({ id: 1, text: 'test' }).subscribe({
      error: done,
      complete: () => {
        try {
          expect(client.index.mock.calls.length).toBe(0);
          done();
        } catch (e) {
          done(e);
        }
      },
    });
  });

  it('should be called with correct id when using custom selector', (done) => {
    loader = new ElasticLoader(
      { nodes: ['http://localhost:9200'] },
      'testIndex',
      () => true,
      (o) => o.myId
    );
    (loader as any).esClient = client;

    loader.write({ myId: 1, text: 'test' }).subscribe({
      error: done,
      complete: () => {
        try {
          expect(client.index.mock.calls.length).toBe(1);
          expect(client.index.mock.calls[0][0]).toMatchSnapshot();
          done();
        } catch (e) {
          done(e);
        }
      },
    });
  });

  it('should reject when no id is provided', (done) => {
    loader = new ElasticLoader({ nodes: ['http://localhost:9200'] }, 'testIndex');
    (loader as any).esClient = client;

    loader.write({ myId: 1, text: 'test' }).subscribe({
      error: () => {
        done();
      },
      complete: () => {
        done(new Error('did not throw.'));
      },
    });
  });
});
