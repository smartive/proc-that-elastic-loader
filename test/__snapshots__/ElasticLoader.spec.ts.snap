// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElasticLoader should be called with correct id when using custom selector 1`] = `
{
  "body": {
    "myId": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
}
`;

exports[`ElasticLoader should use correct body 1`] = `
{
  "body": {
    "id": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
}
`;

exports[`ElasticLoader should use correct id 1`] = `
{
  "body": {
    "id": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
}
`;

exports[`ElasticLoader should use correct index 1`] = `
{
  "body": {
    "id": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
}
`;

exports[`ElasticLoader should use correct type 1`] = `
{
  "body": {
    "id": 1,
    "text": "test",
  },
  "id": 1,
  "index": "testIndex",
}
`;
